DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/src/gzip/gzip_decompressor.hpp

```diff
@@ -0,0 +1,307 @@
+  // check that the input and output pipe types are actually pipes
+  static_assert(fpga_tools::is_sycl_pipe_v<InPipe>);
+  static_assert(fpga_tools::is_sycl_pipe_v<OutPipe>);
+
+  // 'literals_per_cycle' must be greater than 0 and a power of 2
+  static_assert(literals_per_cycle > 0);
+  static_assert(fpga_tools::IsPow2(literals_per_cycle));
+
+  // the inter-kernel pipes for the GZIP decompression engine
+  using GzipMetadataToHuffmanPipe =
+      sycl::ext::intel::pipe<GzipMetadataToHuffmanPipeID,
+                             FlagBundle<ByteSet<1>>>;
+  using HuffmanToLZ77Pipe =
+      sycl::ext::intel::pipe<HuffmanToLZ77PipeID,
+                             FlagBundle<GzipLZ77InputData>,
+                             kHuffmanToLZ77PipeDepth>;
+
+  // submit the GZIP decompression kernels
+  auto header_event =
+      SubmitGzipMetadataReader<GzipMetadataReaderKernelID, InPipe,
+                               GzipMetadataToHuffmanPipe>(
+          q, in_count, hdr_data_out, crc_out, count_out);
+  auto huffman_event =
+      SubmitHuffmanDecoder<HuffmanDecoderKernelID, GzipMetadataToHuffmanPipe,
+                           HuffmanToLZ77Pipe>(q);
+
+  // the design only needs a ByteStacker kernel when literals_per_cycle > 1
+  if constexpr (literals_per_cycle > 1) {
+    using LZ77ToByteStackerPipe =
+        sycl::ext::intel::pipe<LZ77ToByteStackerPipeID,
+                               FlagBundle<BytePack<literals_per_cycle>>>;
+
+    auto lz77_event = SubmitLZ77Decoder<LZ77DecoderKernelID, HuffmanToLZ77Pipe,
+                                        LZ77ToByteStackerPipe,
+                                        literals_per_cycle,
+                                        kGzipMaxLZ77Distance,
+                                        kGzipMaxLZ77Length>(q);
+    auto byte_stacker_event =
+        SubmitByteStacker<ByteStackerKernelID, LZ77ToByteStackerPipe, OutPipe,
+                          literals_per_cycle>(q);
+
+    return {header_event, huffman_event, lz77_event, byte_stacker_event};
+  } else {
+    auto lz77_event =
+        SubmitLZ77Decoder<LZ77DecoderKernelID, HuffmanToLZ77Pipe, OutPipe,
+                          literals_per_cycle, kGzipMaxLZ77Distance,
+                          kGzipMaxLZ77Length>(q);
+    return {header_event, huffman_event, lz77_event};
+  }
+}
+
+// declare kernel and pipe names at the global scope to reduce name mangling
+class ProducerId;
+class ConsumerId;
+class InPipeId;
+class OutPipeId;
+
+// the input and output pipe
+using InPipe = sycl::ext::intel::pipe<InPipeId, ByteSet<1>>;
+using OutPipe =
+    sycl::ext::intel::pipe<OutPipeId, FlagBundle<BytePack<kLiteralsPerCycle>>>;
+
+//
+// The GZIP decompressor. See ../common/common.hpp for more information.
+//
+template <unsigned literals_per_cycle>
+class GzipDecompressor : public DecompressorBase {
+ public:
+  std::optional<std::vector<unsigned char>> DecompressBytes(
+      sycl::queue &q, std::vector<unsigned char> in_bytes, int runs,

```

Comment:
Should this be a vector reference or do you want the compiler to pass in a copy?